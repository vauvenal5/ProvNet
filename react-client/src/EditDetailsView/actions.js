import ContractDetails from "../models/ContractDetails";

export const types = {
    editDetails: "EDIT_DETAILS",
    nop: "NO_OPERATION"
}

export const onEditDetails = (address, title, desc, url) => ({
    type: types.editDetails,
    address: address,
    details: new ContractDetails(title, desc, url)
});

export const onNop = () => ({
    type: types.nop
});