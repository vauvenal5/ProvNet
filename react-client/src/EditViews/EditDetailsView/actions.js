import { ContractDetails } from "./imports";

export const types = {
    editDetails: "EDIT_DETAILS",
    editDetailsError: "EDIT_DETAILS_ERROR",
    editDetailsSuccess: "EDIT_DETAILS_SUCCESS",
    editDetailsModalClear: "EDIT_DETAILS_MODAL_CLEAR",
    nop: "NO_OPERATION"
}

export const onEditDetails = (address, title, desc, url) => ({
    type: types.editDetails,
    address: address,
    details: new ContractDetails(title, desc, url)
});

export const onEditDetailsError = (address, error) => ({
    type: types.editDetailsError,
    address: address,
    error: error
});

export const onEditDetailsSuccess = (address) => ({
    type: types.editDetailsSuccess,
    address: address
})

export const onEditDetailsModalClear = (address) => ({
    type: types.editDetailsModalClear,
    address: address
})

export const onNop = () => ({
    type: types.nop
});