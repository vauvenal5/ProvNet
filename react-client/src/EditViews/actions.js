import { Tag } from "./imports";
import EditModelSelector from "../models/selectors/EditModelSelector";
import { ContractDetails } from "../models";

export const types = {
    nop: "NO_OPERATION",

    editModalOpen: "EDIT_MODAL_OPEN",
    editError: "EDIT_ERROR",
    editSuccess: "EDIT_SUCCESS",
    editModalClear: "EDIT_MODAL_CLEAR",

    editTag: "EDIT_TAG",
    editDetails: "EDIT_DETAILS",
    addUser: "ADD_USER",
    
    deployContract: "DEPLOY_CONTRACT",
    deployContractSuccess: "DEPLOY_CONTRACT_SUCCESS",
    deployContractModalOpen: "DEPLOY_CONTRACT_MODAL_OPEN"
}

// edit

const onEditBase = (type, address, id, payload) => ({
    type: type,
    address: address,
    id: id,
    payload: payload
})

export const onDeployContract = (payload) => onEditBase(types.deployContract, EditModelSelector.newContractKey, EditModelSelector.newContractKey, payload);

export const onEditTag = (address, id, tag, origTag) => onEditBase(types.editTag, address, id, {
    tag,
    origTag
});

export const onAddUser = (address, user, origUser) => onEditBase(
    types.addUser,
    address,
    EditModelSelector.userKey,
    {
        user,
        origUser
    }
)

export const onEditDetails = (address, title, desc, url, origDetails) => onEditBase(
    types.editDetails, 
    address, 
    EditModelSelector.detailsKey, 
    {
        details: new ContractDetails(title, desc, url),
        origDetails
    }
);

// other

export const onNop = () => ({
    type: types.nop
});

// open

const onEditModalOpenBase = (type, value, address, id) => ({
    type: type,
    value: value,
    address: address,
    id: id
});

export const onDeployContractOpen = (value) => onEditModalOpenBase(types.deployContractModalOpen, value, EditModelSelector.newContractKey, EditModelSelector.newContractKey);

export const onEditModalOpen = (value, address, id) => onEditModalOpenBase(types.editModalOpen, value, address, id);

export const onEditDetailsModalOpen = (value, address) => onEditModalOpen(value, address, EditModelSelector.detailsKey); 

// clear

export const onEditModalClear = (address, id) => ({
    type: types.editModalClear,
    address: address,
    id: id
})

// success

export const onEditSuccessBase = (type, address, id, payload) => ({
    type: type,
    address: address,
    id: id,
    payload: payload
})

export const onEditSuccess = (address, id) => onEditSuccessBase(types.editSuccess, address, id);

export const onEditDetailsSuccess = (address) => onEditSuccess(
    address, 
    EditModelSelector.detailsKey
);

export const onDeployContractSuccess = (payload) => onEditSuccessBase(types.deployContractSuccess, EditModelSelector.newContractKey, EditModelSelector.newContractKey, payload);

// error

export const onEditError = (address, id, error) => ({
    type: types.editError,
    address: address,
    id: id,
    error: error
})

export const onDeployContractError = (error) => onEditError(EditModelSelector.newContractKey, EditModelSelector.newContractKey, error);

export const onEditDetailsError = (address, error) => onEditError(address, EditModelSelector.detailsKey, error);