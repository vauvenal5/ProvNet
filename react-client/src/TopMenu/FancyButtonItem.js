import React from 'react';
import { Menu, Icon, Dimmer, Loader } from 'semantic-ui-react';
import EditModel from '../models/EditModel';
import { ErrorIcon } from '../DefaultIcons/ErrorIcon';
import { SuccessIcon } from '../DefaultIcons/SuccessIcon';

export const FancyButtonItem = ({onClick, editModel = new EditModel(), icon, childIcon, disabled}) => {
    return (
        <Menu.Item onClick={() => onClick()} disabled={disabled}>
            <Dimmer active={editModel.isLoading()}>
                <Loader />
            </Dimmer>
            <Dimmer active={editModel.isSuccess()}>
                <SuccessIcon size="big"/>
            </Dimmer>
            <Dimmer active={editModel.isError()}>
                <ErrorIcon size="big"/>
                {/* <Icon name="warning" color="red" size="big" /> */}
            </Dimmer>
            <Icon.Group size="big">
                <Icon name={icon}/>
                <Icon name={childIcon} color="blue" corner disabled={disabled}/>
            </Icon.Group>
        </Menu.Item>
    );
};