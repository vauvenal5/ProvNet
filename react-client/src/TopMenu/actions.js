export const types = {
    topMenuSelect: "TOP_MENU_SELECT",
    searchAddress: "SEARCH_ADDRESS",
};

export const onClick = (name) => ({
    type: types.topMenuSelect,
    name: name,
});

export const onSearchAddress = (address) => ({
    type: types.searchAddress,
    address: address,
});