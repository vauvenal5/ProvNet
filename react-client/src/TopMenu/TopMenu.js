import React from 'react';
import { connect } from 'react-redux';

import { Menu, Icon, Form } from 'semantic-ui-react';

import * as actions from './actions';

import logo from './logo.svg';

export const TopMenu = ({address = "", searchSubmit}) => {

    const handleChange = (e, { name, value }) => {
        address = value;
    };

    const onSubmit = () => searchSubmit(address);

    return (
        <Menu stackable>
            <Menu.Item>
                <img src={logo} />
            </Menu.Item>

            <Menu.Menu position="right">
                <Menu.Item>
                    <Form onSubmit={onSubmit}>
                        <Form.Input 
                        name="address" 
                        onChange={handleChange}
                        icon={
                            <Icon name='search' inverted circular link onClick={onSubmit} />
                        } 
                        placeholder='Contract Address...'>
                        </Form.Input>
                    </Form>
                </Menu.Item>
            </Menu.Menu>
        </Menu>
    );
}

//container part
export const mapStateToProps = (state) => {
    return {
        address: state["topMenu"].address,
    };
}

export const mapDispatchToProps = (dispatch) => {
    return {
        searchSubmit: (address) => dispatch(actions.onSearchAddress(address)),
    };
}

export default connect(mapStateToProps, mapDispatchToProps)(TopMenu);