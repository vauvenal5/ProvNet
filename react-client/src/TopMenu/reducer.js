import * as actions from "./actions";
import * as modelActions from "../modelActions";
import { ofType, combineEpics } from "redux-observable";
import { map, withLatestFrom, flatMap, filter } from 'rxjs/operators';
import { of } from 'rxjs';

const searchAddressEpic = (action$) => action$.pipe(
    ofType(actions.types.searchAddress),
    //todo-sv: check if contract exists
    flatMap(action => of(
        modelActions.onContractLoad(action.address)
    )),
);

const contractLoadedEpic = (action$, state$) => action$.pipe(
    ofType(modelActions.types.contractDetailsLoaded),
    withLatestFrom(state$),
    filter(([action, state]) => 
        action.address === state.topMenu.address
    ),
    map(([action]) => modelActions.onContractSelect(action.address)),
)

export const epic = combineEpics(
    searchAddressEpic,
    contractLoadedEpic,
);

export const reducer = (state={activeItem: ""}, action) => {
    switch(action.type) {
        case actions.types.searchAddress:
            return Object.assign({}, state, {
                address: action.address,
            });
        default:
            return state;
    }
}