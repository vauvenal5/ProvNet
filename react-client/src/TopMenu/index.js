import Component from "./TopMenu";
import {reducer, selected, epic} from "./reducer";

export default {Component, reducer, selected, epic};