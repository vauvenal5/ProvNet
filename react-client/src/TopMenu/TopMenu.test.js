import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { mapDispatchToProps, TopMenu, mapStateToProps } from './TopMenu';
import * as actions from './actions';
import testContract from "../models/ProvContract.testHelper";
import { Form, Icon } from 'semantic-ui-react';

Enzyme.configure({ adapter: new Adapter() });

const setup = (propOverrides) => {
    const props = Object.assign({}, propOverrides);

    const wrapper = shallow(<TopMenu {...props}/>);

    return {
        props,
        wrapper,
        expected: wrapper.debug()
    }
}

describe("TopMenu", () => {
    it("should render empty view", () => {
        const {expected} = setup();
        expect(expected).toMatchSnapshot();
    });

    it("should map correct state to props", () => {
        const state = {
            topMenu: {
                address: testContract.address
            }
        };

        expect(mapStateToProps(state)).toEqual(state.topMenu);
    })

    it("should call correct dispatch on item click", (done) => {
        const dispatch = (action) => {
            expect(action).toEqual(actions.onSearchAddress(testContract.address));
            done();
        };
        mapDispatchToProps(dispatch).searchSubmit(testContract.address);
    });

    it("should submit search with correct address", (done) => {
        const { wrapper } = setup({searchSubmit: (address) => {
            expect(address).toEqual(testContract.address);
            done();
        }})

        wrapper.find(Form.Input).prop("onChange")("event", { 
            name: "address",
            value: testContract.address
        });

        wrapper.find(Form).simulate("submit");
    });
});Icon