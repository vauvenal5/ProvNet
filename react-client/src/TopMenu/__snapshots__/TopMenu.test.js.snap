// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TopMenu should render empty view 1`] = `
"<Menu stackable={true}>
  <MenuItem>
    <img src=\\"logo.svg\\" />
  </MenuItem>
  <MenuMenu position=\\"right\\">
    <MenuItem>
      <Form onSubmit={[Function: onSubmit]} as=\\"form\\">
        <FormInput name=\\"address\\" onChange={[Function: handleChange]} icon={{...}} placeholder=\\"Contract Address...\\" as={[Function: FormField]} control={[Function: Input]} />
      </Form>
    </MenuItem>
  </MenuMenu>
</Menu>"
`;
